### ISSUE TITLE
Rosette Summary file name

### ISSUE THREAD

**Issue by upsonp** [pyoceans/python-ctd#204]
I've come across an issue with yet again how DFO does things. I'm processing BTL/ROS files from what we call a 'fix station' and the files have next to no metadata. The BTL file processes fine, but when I try to use ctd.read.from_summary() I get an error on line 451 `name = _basename(fname)[1]`

The files don't have a `* FileName = xxxx` line so there is no file name.

The ctd.read.from_btl() function solves this on lines 278-280 with:
```
    if "name" not in metadata:
        name = _basename(fname)[1]
        metadata["name"] = str(name)
```

I'd like to do the same thing for the ctd.read.from_cnv() function, called by the ctd.read.from_summary() function if it's not an issue.

**Comment by ocefpaf** [comment]
> I'd like to do the same thing for the ctd.read.from_cnv() function, called by the ctd.read.from_summary() function if it's not an issue.

Yes please. We should have that kind of fallback in all read functions.

**Comment by upsonp** [comment]
It's a bit more complicated to fix the issue for the ctd.read.from_edf() function, the only other function to call ctd.read._basename().

The real issue is we're mixing passing a file path and an io.StringIO object to each of the reader functions, but then later we just make the assumption a file path was passed in and then call _basename() to get the path, name and extension.

I've fixed my issue and included a fixstation_hl_02.ros file with a unit test using the solution I described above, but maybe a better solution would be to rewrite the ctd.read._basename() function.

currently the function looks like:

```python
def _basename(fname):
    """Return file name without path."""
    if not isinstance(fname, Path):
        fname = Path(fname)
    path, name, ext = fname.parent, fname.stem, fname.suffix
    return path, name, ext
```
And when an io.StringIO object is passed in the `fname = Path(fname)` line throws an exception booting us back out to wherever the _basename call came from.

We can solve this in a couple of ways.
1) we rewrite the _basename function:
```python
def _basename(fname):
    """Return file name without path."""
    if isinstance(fname, io.StringIO):
        return 'unknown', 'unknown', 'unknown'

    if not isinstance(fname, Path):
        fname = Path(fname)
    path, name, ext = fname.parent, fname.stem, fname.suffix
    return path, name, ext
```

2) we can use a try-except block wherever _basename is called that if for *any reason* an exception is thrown we handle the issue specific to the function calling _basename.

**Comment by upsonp** [comment]
So something like:

```python
def _basename(fname):
    """Return file name without path."""
    if isinstance(fname, io.StringIO):
        return 'unknown', str(hash(fname)), 'unknown'

    if not isinstance(fname, Path):
        fname = Path(fname)
    path, name, ext = fname.parent, fname.stem, fname.suffix
    return path, name, ext
```

**Comment by ocefpaf** [comment]
Looks good. I would restrict the hash to few characters though. Maybe something like:

```python
import hashlib
hashlib.shake_256(fname).hexdigest(5)
```

**Comment by ocefpaf** [comment]
BTW, you don't have to tackle that here. If you are OK with the solution for your case we can open an issue for this and solve it in another PR.

**Comment by upsonp** [comment]
That would be preferable. 😄

**Comment by ocefpaf** [comment]
> That would be preferable. 😄

Let's leave this issue open though b/c we want to solve that at some point. I merged your PR and will mint a release soon. Thanks!
